src/member-api.js

```diff
@@ -54,42 +77,140 @@ export class MemberApi extends TypedEmitter {
+    const cleanup = () => {
+      for (const fn of cleanupFns) {
+        try {
+          fn()
+        } catch (_err) {
+          console.error('Cleanup function failed')
+        }
+      }
+    }
 
-    // since we are always getting #ownDeviceId,
-    // this should never throw (see comment on getById), but it pleases ts
-    if (!deviceName)
-      throw new Error(
-        'Internal error trying to read own device name for this invite'
-      )
+    try {
+      let handleAborted = noop
+      const timeoutId = setTimeout(() => {
+        handleAborted = () => {
+          throw new Error('Invite timed out')
+        }
+      }, timeout)
+      cleanupFns.push(() => {
+        clearTimeout(timeoutId)
+      })
 
-    const projectId = projectKeyToId(this.#projectKey)
-    const project = await this.#dataTypes.project.getByDocId(projectId)
+      this.#deviceIdsWithPendingInvites.add(deviceId)
+      cleanupFns.push(() => {
+        this.#deviceIdsWithPendingInvites.delete(deviceId)
+      })
 
-    if (!project.name)
-      throw new Error('Project must have a name to invite people')
+      const { name: invitorName } = await this.getById(this.#ownDeviceId)
+      // since we are always getting #ownDeviceId,
+      // this should never throw (see comment on getById), but it pleases ts
+      if (!invitorName)
+        throw new Error(
+          'Internal error trying to read own device name for this invite'
+        )
+      handleAborted()
 
-    const response = await this.#rpc.invite(deviceId, {
-      projectKey: this.#projectKey,
-      encryptionKeys: this.#encryptionKeys,
-      projectInfo: { name: project.name },
-      roleName: roleName || ROLES[roleId].name,
-      roleDescription,
-      invitorName: deviceName,
-      timeout,
-    })
+      const projectId = projectKeyToId(this.#projectKey)
+      const projectPublicId = projectKeyToPublicId(this.#projectKey)
+      const project = await this.#dataTypes.project.getByDocId(projectId)
+      const projectName = project.name
+      if (!projectName)
+        throw new Error('Project must have a name to invite people')
+      handleAborted()
 
-    if (response === InviteResponse_Decision.ACCEPT) {
-      await this.#roles.assignRole(deviceId, roleId)
-    }
+      const inviteId = crypto.randomBytes(32)
+
+      const receiveInviteAbortController = new AbortController()
+      cleanupFns.push(() => {
+        receiveInviteAbortController.abort()
+      })
+      const receiveInviteResponsePromise =
+        /** @type {typeof onceSatisfied<TypedEmitter<import('./local-peers.js').LocalPeersEvents>, 'invite-response'>} */ (
+          onceSatisfied
+        )(
+          this.#rpc,
+          'invite-response',
+          (peerId, inviteResponse) =>
+            timingSafeEqual(peerId, deviceId) &&
+            timingSafeEqual(inviteId, inviteResponse.inviteId),
+          { signal: receiveInviteAbortController.signal }
+        ).then((args) => args?.[1])
+
```

Comment:
```suggestion

receiveInviteResponsePromise.catch(noop)
```

Reply:
See [my comment above](https://github.com/digidem/mapeo-core-next/pull/484#discussion_r1520359400)—I don't think this promise will ever reject.

Reply:
It now might reject! Updated in 759e5c5cf90daf00d346c34df35ec3768420b4e4.